robottelo/ssh.py

```diff
@@ -1,285 +1,95 @@
-
-    def run(self, cmd, *args, **kwargs):
-        """This method exists to allow the reuse of existing connections when
-        running multiple ssh commands as in the following example of use:
-
-            with robottelo.ssh.get_connection() as connection:
-                connection.run('ls /tmp')
-                connection.run('another command')
-
-        `self` is always passed as the connection when used in context manager
-        only when using `ssh.get_connection` function.
-
-        Note: This method is named `run` to avoid conflicts with existing
-        `exec_command` and local function `execute_command`.
-        """
-        return execute_command(cmd, self, *args, **kwargs)
-
-
-def _call_paramiko_sshclient():  # pragma: no cover
-    """Call ``paramiko.SSHClient``.
-
-    This function does not alter the behaviour of ``paramiko.SSHClient``. It
-    exists solely for the sake of easing unit testing: it can be overridden for
-    mocking purposes.
-
-    """
-    return SSHClient()
 
 
+@lru_cache
 def get_client(
     hostname=None,
     username=None,
     password=None,
-    key_filename=None,
-    key_string=None,
     timeout=None,
     port=22,
 ):
-    """Returns a SSH client connected to given hostname
+    """Returns a host object that provides an ssh connection
 
     Processes ssh credentials in the order: password, key_filename, ssh_key
     Config validation enforces one of the three must be set in settings.server
     """
-    # temporary workaround due to import order.
     from robottelo.config import settings
+    from robottelo.hosts import Host
 
-    hostname = hostname or settings.server.hostname
-    username = username or settings.server.ssh_username
-    password = password or settings.server.ssh_password
-    if password is None:
-        key_filename = key_filename or settings.server.ssh_key
-    if password is None and key_filename is None:
-        key_string = key_string or settings.server.ssh_key_string
-        key_string = paramiko.rsakey.RSAKey.from_private_key(StringIO(str(key_string)))
-    timeout = timeout or settings.server.ssh_client.connection_timeout
-    client = _call_paramiko_sshclient()
-    client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
-    client.connect(
-        hostname=hostname,
-        username=username,
-        key_filename=key_filename,
-        pkey=key_string,
-        password=password,
-        timeout=timeout,
-        port=port,
-        allow_agent=False if password else True,
-        look_for_keys=False if password else True,
+    client = Host(
+        hostname=hostname or settings.server.hostname,
+        username=username or settings.server.ssh_username,
+        password=password or settings.server.ssh_password,
     )
-    client._id = hex(id(client))
+    if timeout:
+        client.default_timeout = timeout
     return client
 
 
-@contextmanager
-def get_connection(
+def add_authorized_key(
```

Comment:
This was moved to the Hosts module in the initial commit to the `unstable` branch:

https://github.com/SatelliteQE/robottelo/blob/unstable/robottelo/hosts.py#L424